core/src/main/java/org/apache/brooklyn/core/config/BasicConfigInheritance.java

```diff
@@ -28,51 +32,188 @@
-     * subclasses may pass null if they provide a custom implementaton of {@link #resolveWithParentCustomStrategy(ConfigValueAtContainer, ConfigValueAtContainer, org.apache.brooklyn.config.ConfigInheritance.ConfigInheritanceContext)} */
+     * subclasses may pass null or a different string if they provide a custom implementaton 
+     * of {@link #resolveWithParentCustomStrategy(ConfigValueAtContainer, ConfigValueAtContainer, org.apache.brooklyn.config.ConfigInheritance.ConfigInheritanceContext)} */
     @Nullable protected final String conflictResolutionStrategy;
-    /** use-local-default-value? true/false; if true, overwrite above means "always ignore (even if null)"; default false
-     * whereas merge means "use local default (if non-null)" */
-    protected final boolean useLocalDefaultValue;
+    
+    /** @deprecated since 0.10.0 when this was introduced, now renamed {@link #localDefaultResolvesWithAncestorValue} */
+    @Deprecated protected final Boolean useLocalDefaultValue;
+    /** whether a local default value should be considered for resolution in the presence of an ancestor value.
+     * can use true with overwrite to mean don't inherit, or true with merge to mean local default merged on top of inherited
+     * (but be careful here, if local default is null in a merge it will delete ancestor values).
+     * <p>
+     * in most cases this is false, meaning a default value is ignored if the parent has a value.
+     * <p>
+     * null should not be used. a boxed object is taken (as opposed to a primitive boolean) only in order to support migration.    
+     */
+    @Nonnull
+    protected final Boolean localDefaultResolvesWithAncestorValue;
 
-    protected BasicConfigInheritance(boolean isReinherited, @Nullable String conflictResolutionStrategy, boolean useLocalDefaultValue) {
+    /** whether a default set in an ancestor container's key definition will be considered as the
+     * local default value at descendants who don't define any other value (nothing set locally and local default is null);
+     * <p>
+     * if true (now the usual behaviour), if an ancestor defines a default and a descendant doesn't, the ancestor's value will be taken as a default.
+     * if it is also the case that localDefaultResolvesWithAncestorValue is true at the <i>ancestor</i> then a descendant who
+     * defines a local default value (with this field true) will have its conflict resolution strategy
+     * applied with the ancestor's default value.
+     * <p>
+     * if this is false, ancestor defaults are completely ignored; prior to 0.10.0 this was the normal behaviour,
+     * but it caused surprises where default values in parameters did not take effect.
+     * <p>
+     * null should not be used. a boxed object is taken (as opposed to a primitive boolean) only in order to support migration.    
+     */
+    @Nonnull
+    protected final Boolean ancestorDefaultInheritable;
```

Comment:
have thought about this:  the two conditions are independent dimensions, and it would matter if we used `localDefaultResolvesWithAncestorValue` for any purpose other than blocking inheritance from ancestors.  have added more comments to explain, including your words about the "independent dimensions" which is a good point:  including the control dimensions lets the logic be correct, even if we only expose a subset of possibilities to the end user (which simplifies a little bit the complexity of config inheritance).